src/lidar/read_rp_lidar.cpp

```diff
@@ -12,60 +12,84 @@
 std::thread lidar_thread;
-points_t last_points = {};
+DataPoint<points_t> last_points = {};
 std::mutex points_lock;
 
 namespace lidar {
 
-bool initializeLidar() {
-    if (!lidar_initialized) {
-        if (!rp_lidar.checkHealth()) {
-            perror("failed to connect to rp lidar");
-        }
-        rp_lidar.setMaxDistance(16.0);
-        lidar_thread = std::thread(&readLidarLoop);
-    }
-    return lidar_initialized;
-}
-
+/**
+ * @brief RP Lidar continuously scans environment, updates current timeframe with lidar data
+ */
 void readLidarLoop() {
     using namespace std::chrono;
     std::vector<Polar2D> polarPts;    
     while (true) {
         auto scan = rp_lidar.poll();
         if (auto scan = rp_lidar.poll()) {
-            // Converts data into Polar Coord
-            std::vector<Polar2D> currFrames;
+            std::vector<point_t> pts;
             double dtheta = (scan.value().angle_max-scan.value().angle_min)/(scan.value().ranges.size()-1);
             for (unsigned long i = 0; i < scan.value().ranges.size(); i++) {
                 double rad = dtheta*i;
-                double dist = scan.value().ranges[i] * 1000;
+                double dist = scan.value().ranges[i] * MM_TO_M;
 
                 Polar2D frame{dist, rad};
-                currFrames.push_back(frame);
+                pts.push_back(lidar::polarToCartesian2(frame));
             }
         
-            // Converts to regular points
-            std::vector<point_t> pts(currFrames.size());
-            point_t origin({0, 0, 1});
-            for (unsigned long i = 0; i < currFrames.size(); i++) {
-                pts[i] = lidar::polarToCartesian2(currFrames[i]);
-            }
             points_lock.lock();
-            last_points = pts;
+            datatime_t time = lidar_time;
+            last_points = {time, pts};
             points_lock.unlock();
-            lidar_initialized = true;
         } else {
             perror("failed to get frame");
+            points_lock.lock();
+            datatime_t time = lidar_time;
+            last_points = {};
+            points_lock.unlock();
             continue;
         }
     } 
 }
 
+/**
+ * @brief Startsup RPLidar with default settings
+ */
+bool initializeLidar() {
+    if (!lidar_initialized) {
+        if (!rp_lidar.checkHealth()) {
+            perror("failed to connect to rp lidar");
+        } else {
+            rp_lidar.setMaxDistance(16.0);
+            points_lock.lock();
+            lidar_thread = std::thread(&readLidarLoop);
+            lidar_initialized = true;
+            points_lock.unlock();
+        }
+    }
+    return lidar_initialized;
+}
+
+/**
+ * @brief Provides RP Lidar data at current timeframe
+ */
+DataPoint<points_t> readLidar() {
+    if (lidar_initialized) {
+        DataPoint<points_t> pts;
+        points_lock.lock();
+        pts = last_points;
+        points_lock.unlock();
+        return pts;
+    }
+    return {};
+}
```

Comment:
If you make the change suggested above wrt `initializeLidar()` then `points_lock` should be locked before the if statement to avoid race conditions.

If it helps you could use the scoped [`lock_guard`](https://en.cppreference.com/w/cpp/thread/lock_guard) instead of manually calling `lock()` and `unlock()`, but this way is fine too.